chapters/optimization/index.md

```diff
@@ -0,0 +1,145 @@
+## Flow Optmization
+
+##### Flow Optimization eases the process of rerunning Flows with different parameter sets. It allows out of the box hyperparameter tuning inside Jina.
+
+### Flow Optimization is hyperparameter tuning on steroids
+
+A common principle when building a Flow in Jina is:
+
+1. Design the high level `flow.yml`
+2. Define several `pods.yml` files for all needed Executors
+3. Repeat until happy with results:
+  a. Change variable values in different Executors (e.g. used model, used model layer, details of the segmenter)
+  b. Index some data
+  c. Query some data and look at the results
+
+Step 3 is time consuming and often the user has no idea, which values to test.
+Flow Optimization automates exactly these steps.
+It can be done via python code or `.yml` definitions as commonly used around Jina.
+
+### Before you start
+
+Make sure you install Jina via `Installation <https://docs.jina.ai/chapters/install/os/index.html>`_ with `jina[optimizer]`.
+
+### Using FlowOptimization
+
+In this toy example, we choose the optimal layer of an encoder for the final embedding.
+This is a common practice in ML.
+The best semantic representation for a given problem might not be the last layer of a given model.
+
+Flow Optimization requires the following components:
+
+- At least one Flow and the corresponding Pod definitions via JAML
+- An Evaluator Executor in at least one of the Flows
+- Documents, which are send to each Flow
+- A `parameter.yml` file describing the optimization scenario
+- A `FlowRunner` object, which allows repeatedly running the same Flow with different configurations.
+
+Let's define a `flow.yml`:
+
+```yaml
+!Flow
+version: '1'
+env:
+  JINA_ENCODER_LAYER_VAR: ${{JINA_ENCODER_LAYER}}
+pods:
+  - uses: encoder.yml
+  - uses: EuclideanEvaluator
+```
+
+The `EuclideanEvaluator` is used for calculating the distance between the encoded and an expected vector.
+Furthermore, we need the corresponding `encoder.yml`:
+
+```yaml
+!SimpleEncoder
+with:
+  index: ${{JINA_ENCODER_LAYER_VAR}}
+```
+
+```python
+class SimpleEncoder(BaseEncoder):
+
+    ENCODE_LOOKUP = {
+        'cat': [1, 3, 5],
+        'pokemon': [2, 4, 7],
+        'dog': [0, 2, 5],
+    }
+
+    def __init__(self, layer=0, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._layer = layer
+
+    def encode(self, data: 'np.ndarray', *args, **kwargs) -> 'np.ndarray':
+        return np.array([[self.ENCODE_LOOKUP[data[0]][self._layer]]])
```

Comment:
`self.ENCODE_LOOKUP` is a dict where the keys are strings and `data` is a `np.ndarray`.

It seems a bit odd that `data[0]` is used as a key for `ENCODE_LOOKUP ` because  `data` is a `np.ndarray`.

 It is true that `np.ndarray` migtht have as `dtype` a string but that is a very uncommon use. If that is the case I might write it explicitly.

If the input is not necessarily a `np.ndarray` object, it could be written as `Iterable[str]` which would consider the case of `np.ndarray` of  strings.
```
def encode(self, data:`Iterable[str]`, *args, **kwargs) -> 'np.ndarray':
         return np.array([[self.ENCODE_LOOKUP[data[0]][self._layer]]])
```

Reply:
Great point. Apparently `data: Sequence[str]`, otherwise we couldn't check and index.